ovirt/resource_ovirt_affinity_group.go

```diff
@@ -0,0 +1,425 @@
+	d.Set("vm_enabled", affinityGroup.MustVmsRule().MustEnabled())
+	d.Set("vm_enforcing", affinityGroup.MustVmsRule().MustEnforcing())
+	d.Set("vm_positive", affinityGroup.MustVmsRule().MustPositive())
+
+	hosts := affinityGroup.MustHosts().Slice()
+	hostNames := make([]string, len(hosts))
+	for i, h := range hosts {
+		hostNames[i] = h.MustId()
+	}
+	sort.Strings(hostNames)
+	d.Set("host_list", hostNames)
+
+	vms := affinityGroup.MustVms().Slice()
+	vmNames := make([]string, len(vms))
+	for i, v := range vms {
+		vmNames[i] = v.MustId()
+	}
+	sort.Strings(vmNames)
+	d.Set("vm_list", vmNames)
+
+	return nil
+}
+
+func resourceOvirtAffinityGroupUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*ovirtsdk4.Connection)
+
+	group := ovirtsdk4.NewAffinityGroupBuilder()
+	attributeUpdate := false
+
+	if d.HasChange("name") {
+		group.Name(d.Get("name").(string))
+		attributeUpdate = true
+	}
+	if d.HasChange("description") {
+		group.Description(d.Get("description").(string))
+		attributeUpdate = true
+	}
+	if d.HasChange("cluster_id") {
+		group.Cluster(
+			ovirtsdk4.NewClusterBuilder().
+				Id(d.Get("cluster_id").(string)).
+				MustBuild())
+		attributeUpdate = true
+	}
+
+	vmRuleBuilder := ovirtsdk4.NewAffinityRuleBuilder()
+	vmRuleUpdate := false
+	if d.HasChange("vm_positive") {
+		vmRuleBuilder.Positive(d.Get("vm_positive").(bool))
+		vmRuleUpdate = true
+	}
+	if d.HasChange("vm_enforcing") {
+		vmRuleBuilder.Enforcing(d.Get("vm_enforcing").(bool))
+		vmRuleUpdate = true
+	}
+	if d.HasChange("vm_list") {
+		vmRuleBuilder.Enabled(len(d.Get("vm_list").([]interface{})) > 0)
+		vmRuleUpdate = true
+	}
+	if vmRuleUpdate {
+		group.VmsRule(vmRuleBuilder.MustBuild())
+		attributeUpdate = true
+	}
+
+	hostRuleBuilder := ovirtsdk4.NewAffinityRuleBuilder()
+	hostRuleUpdate := false
+	if d.HasChange("host_positive") {
+		hostRuleBuilder.Positive(d.Get("host_positive").(bool))
+		hostRuleUpdate = true
+	}
+	if d.HasChange("host_enforcing") {
+		hostRuleBuilder.Enforcing(d.Get("host_enforcing").(bool))
+		hostRuleUpdate = true
+	}
+	if d.HasChange("host_list") {
+		hostRuleBuilder.Enabled(len(d.Get("host_list").([]interface{})) > 0)
+		hostRuleUpdate = true
+	}
+	if hostRuleUpdate {
+		group.HostsRule(hostRuleBuilder.MustBuild())
+		attributeUpdate = true
+	}
+
+	if attributeUpdate {
+		log.Printf("[DEBUG] Updating %#v", group.MustBuild())
+		_, err := conn.SystemService().
+			ClustersService().
+			ClusterService(d.Get("cluster_id").(string)).
+			AffinityGroupsService().
+			GroupService(d.Id()).
+			Update().
+			Group(group.MustBuild()).
+			Send()
+		if err != nil {
+			log.Printf("[DEBUG] Error updating affinity group (%s): %s", d.Id(), err)
+			return err
+		}
+	}
+
+	if d.HasChange("vm_list") {
+		vmsService := conn.SystemService().
+			ClustersService().
+			ClusterService(d.Get("cluster_id").(string)).
+			AffinityGroupsService().
+			GroupService(d.Id()).
```

Comment:
you should pass here 
conn.SystemService().ClustersService().ClusterService("Cluster ID").AffinityGroupsService().GroupService("VM ID")

Reply:
I don't understand this.

You're suggesting removing the final part of this, so remove ".VmsService()", and updating the signature of updateVmList. 

I don't understand the reasoning behind this - it works as is, and if you remove the above "VmsService()" I would need to add it back in within the updateVmList function?

Reply:
ahh, I'm getting ahead of myself.

Again, this ties into your above comments. If I am disregarding the current list, and simply posting an update, I no longer require the VmsService. That wasn't clear from your comments, but got it now.